File: tests/test_runtime_configure.py
import threading
from pathlib import Path

from simple_pipe.context.frame_meta import FrameInput
from simple_pipe.pipeline.simple_pipeline import SimplePipeline
from simple_pipe.spec.runtime_config import PipelineRuntimeConfig

ROOT = Path(__file__).resolve().parents[1]


def test_configure_while_running_keeps_pipeline_active():
    pipeline = SimplePipeline()
    assert pipeline.build_from_json(ROOT / "configs/demo/graph/five_node.json").ok
    pipeline.configure(PipelineRuntimeConfig.from_dict({"node_params": {"rules": {"min_score": 0.5}}}))

    done = threading.Event()
    pipeline.set_output_callback(lambda _: done.set())
    assert pipeline.start().ok

    assert pipeline.push_frame(FrameInput(image="a", frame_index=0)).ok
    assert pipeline.configure(
        PipelineRuntimeConfig.from_dict({"node_params": {"rules": {"min_score": 0.99}}})
    ).ok
    assert pipeline.is_running()
    assert pipeline.push_frame(FrameInput(image="b", frame_index=1)).ok

    assert done.wait(timeout=3.0)
    pipeline.stop()

File: tests/test_runtime_configure.cpp
#include <atomic>
#include <chrono>
#include <thread>

#include <gtest/gtest.h>

#include "simple_pipe/context/frame_meta.hpp"
#include "simple_pipe/pipeline/simple_pipeline.hpp"
#include "simple_pipe/spec/runtime_config.hpp"
#include "test_helper.hpp"

using namespace simple_pipe;
using namespace simple_pipe::test;

TEST(RuntimeConfigureTest, ConfigureWhileRunningAllowsMorePush) {
  SimplePipeline pipeline;
  ASSERT_TRUE((pipeline.BuildFromJson(GraphPath("five_node"))).IsSuccess());

  nlohmann::json initial = {{"node_params", {{"rules", {{"min_score", 0.5}}}}}};
  ASSERT_TRUE((pipeline.Configure(PipelineRuntimeConfig::FromJson(initial))).IsSuccess());

  std::atomic<int> received{0};
  pipeline.SetOutputCallback([&](const FrameMeta&) { received.fetch_add(1); });
  ASSERT_TRUE((pipeline.StartPipeline()).IsSuccess());

  FrameInput a;
  a.frame_index = 0;
  ASSERT_TRUE((pipeline.PushFrame(a)).IsSuccess());

  nlohmann::json patch = {{"node_params", {{"rules", {{"min_score", 0.99}}}}}};
  ASSERT_TRUE((pipeline.Configure(PipelineRuntimeConfig::FromJson(patch))).IsSuccess());
  EXPECT_TRUE(pipeline.IsRunning());

  FrameInput b;
  b.frame_index = 1;
  ASSERT_TRUE((pipeline.PushFrame(b)).IsSuccess());

  const auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(3);
  while (received.load() < 2 && std::chrono::steady_clock::now() < deadline) {
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }
  pipeline.StopPipeline();
  EXPECT_GE(received.load(), 2);
}
